libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.java

```diff
@@ -65,19 +70,73 @@ public boolean isLastLeg(RouteProgress routeProgress) {
    */
   @Nullable
   public List<Point> calculateRemainingWaypoints(RouteProgress routeProgress) {
-    if (routeProgress.directionsRoute().routeOptions() == null) {
+    RouteOptions routeOptions = routeProgress.directionsRoute().routeOptions();
+    if (routeOptions == null) {
       return null;
     }
-    List<Point> coordinates = new ArrayList<>(routeProgress.directionsRoute().routeOptions().coordinates());
+    List<Point> coordinates = new ArrayList<>(routeOptions.coordinates());
     int coordinatesSize = coordinates.size();
-    int remainingWaypoints = routeProgress.remainingWaypoints();
-    if (coordinatesSize < remainingWaypoints) {
+    int remainingWaypointsCount = routeProgress.remainingWaypointsCount();
+    if (coordinatesSize < remainingWaypointsCount) {
       return null;
     }
-    List<Point> remainingCoordinates = coordinates.subList(coordinatesSize - remainingWaypoints, coordinatesSize);
+    String waypointIndices = routeOptions.waypointIndices();
+    if (waypointIndices == null) {
+      return null;
+    }
+    String[] wayPointIndices = waypointIndices.split(SEMICOLON);
+    String[] remainingWaypointIndices = Arrays.copyOfRange(wayPointIndices,
+      wayPointIndices.length - remainingWaypointsCount, wayPointIndices.length);
+    List<Point> remainingCoordinates = new ArrayList<>();
+    try {
+      int firstRemainingWaypointIndex = Integer.valueOf(remainingWaypointIndices[FIRST_POSITION]);
+      remainingCoordinates = coordinates.subList(firstRemainingWaypointIndex, coordinatesSize);
+    } catch (NumberFormatException ex) {
+      Timber.e("Fail to convert waypoint index to integer");
+    }
     return remainingCoordinates;
   }
 
+  /**
+   * Given a {@link RouteProgress}, this method will calculate the waypoint indices
+   * along the given route based on route option waypoint indices and the progress remaining waypoints coordinates.
+   * Remaining waypoint indices are recalculated based on count of already achieved waypoints.
+   * <p>
+   * If the waypoint indices are empty, this method will return null.
+   *
+   * @param routeProgress for route waypoint indices and remaining coordinates
+   * @return Integer array including the origin waypoint index and the recalculated remaining ones
+   * @since 0.43.0
+   */
+  @Nullable
+  public Integer[] calculateRemainingWaypointIndices(RouteProgress routeProgress) {
```

Comment:
Same here, seems like returning empty array might be a better strategy as opposed to returning `null`?